### PR TITLE
Better support for mli/rei only modules

### PR DESCRIPTION
**Pull request by diml**
Jbuilder now supports them. However, given that we can't check that a module is a pure interfaces and that not respecting this rule could lead to hard to debug errors, the user is required to explicitly list such modules.

The syntax is as follow:

```scheme
(library
 ((name foo)
  (modules_without_implementation (x y z))))
```

### CODE REVIEW

**Comment by jpdeplaix** [comment]
> (modules_with_implementation (x y z))))

Isn't it more ```(modules_without_implementation (x y z))))``` ?

**Comment by diml** [comment]
Indeed... I updated the description

File: src/module_compilation.ml [hunk 1]
```diff
@@ -53,6 +55,12 @@ let build_cm sctx ?sandbox ~dynlink ~flags ~cm_kind ~dep_graphs
           let fn = Option.value_exn (Module.cmt_file m ~dir ml_kind) in
           (fn :: extra_targets, A "-bin-annot")
       in
+      let opaque =
+        if cm_kind = Cmi && not (Module.has_impl m) && ctx.version >= (4, 03, 0) then
+          Arg_spec.A "-opaque"
```
[the remark this answers is not in the document]
Comment (diml) [reply src/module_compilation.ml]: It's to silence a warning from the compiler, for some reason it still looks for the cmx.

> A bit unrelated but others have noted that this is would be desirable in situations where they refer faster recompilation over better generated code.

Indeed, we should add support for that at some point. Is `-opaque` itself enough BTW? Or do we need to *hide* the cmx files as well

**Comment by diml** [comment]
> I noticed that this implementation accepts making the alias module "mli only" when (wrapped true) is set. Is that correct? If it is, what are the semantics of that?

Currerntly the scheme is as follow: for a library named `foo`:
- if the library has no module `Foo`, generate one and use it as alias module
- if the library has a module `Foo`, don't wrap this module and generate the alias module as `Foo__`

There is no point giving an implementation to `Foo__` since the user is not supposed to know it even exist. Now that we support interface only modules, it just seems more natural to make it an interface only module. We might be able to do the same in the first case as well, though I'm not sure it will always work when the user does things such as `let module M = Foo in ...` or `Make(Foo)`.

File: src/module_compilation.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgrinberg) [reply src/module_compilation.ml]: Sure, I've made a ticket for it #497

**Comment by rgrinberg** [comment]
> Now that we support interface only modules, it just seems more natural to make it an interface only module

Yeah, I think that makes sense if it doesn't break anything.

Ok, looks good. Thanks for explaining.

**Comment by diml** [comment]
BTW, I'm wondering if we even need to install foo__.cmi

File: CHANGES.md [hunk 2]
```diff
@@ -26,6 +26,8 @@ next
   directory can't see each other unless one of them depend on the
   other (#472)
 
+- Better support for mli/rei only modules (#490)
```
Comment (samoht) [inline CHANGES.md]: This should be #489

Comment (diml) [reply CHANGES.md]: Indeed, it's now fixed. Thanks

**Pull request merged**